source/nodejs/adaptivecards/src/markdown-formatter.ts

```diff
@@ -1,182 +1,186 @@
-			patternType: 'symmetric',
-			groups: 1,
-		},
-		{
-			type: 'italic',
-            styles:{
+        }, {
+            type: "italic",
+            styles: {
                 start: "<i>",
-                end: "</i>"},
-			pattern: ['_'],
-			patternType: 'symmetric',
-			groups: 1,
-		},
-		{
-			type: 'hyperlink',
+                end: "</i>"
+            },
+            patterns: ['_'],
+            patternType: 'symmetric',
+            groups: 1,
+        }, {
+            type: "hyperlink",
             styles: {
-                html:"<a href='{1}'>{0}</a>"
+                html: "<a href='{1}'>{0}</a>"
             },
-			pattern: ['[]()'],
-			patternType: 'asymmetric',
-			groups: 2,
-		},
+            patterns: ['[]()'],
+            patternType: 'asymmetric',
+            groups: 2,
+        }
     ];
 
-    //combines default config and user config.
-    static setCustomMarkdownRegex(regexArray: any[]){
-        MarkdownFormatter.regexArray = regexArray;
-
-		// prefer user configs
-		for(var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++){
-			for(var j =0; j < MarkdownFormatter.regexArray.length; j++){
-				if(MarkdownFormatter.MD_FORMATTER_CONFIG[i].type == MarkdownFormatter.regexArray[j].type){
-					MarkdownFormatter.MD_FORMATTER_CONFIG[i] = MarkdownFormatter.regexArray[j];
-					MarkdownFormatter.regexArray.splice(j, 1);
-					continue;					
-				}
-			}
-		}
-        MarkdownFormatter.regexArray = MarkdownFormatter.MD_FORMATTER_CONFIG.concat(MarkdownFormatter.regexArray);
-        
+    constructor() {
+        super();
+        this.configuration.forEach(config => {
+            this.initailize(config);
+        });
     }
-    
-
-	constructor(text:string) {
-        this.text = text;
-       
-		for (var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++) {
-			let patternArray = MarkdownFormatter.MD_FORMATTER_CONFIG[i].pattern;
-			let patternType = MarkdownFormatter.MD_FORMATTER_CONFIG[i].patternType;
-            let groups = MarkdownFormatter.MD_FORMATTER_CONFIG[i].groups;
-            let pattern = patternArray[0];
-            
-            if(patternType === "custom"){                
-                pattern = patternArray[0];
+
+    /**
+     * To add custom markdown rules. This will update the default markdown rule is the same type is passed.
+     * 
+     * @param configurations 
+     */
+    public setCustomRules(configurations: IFormatterConfiguration[]) {
+        for (var j = 0; j < configurations.length; j++) {
+            let index = this.configuration.findIndex(config => config.type == configurations[j].type);
+            if (index >= 0) {
+                this.configuration[index] = configurations[j];
+                this.initailize(configurations[j]);
+                continue;
+            } else {
+                this.configuration.push(configurations[j]);
+            }
+        }
+    }
+
+    /**
+     * To add '\' before all the special characters in the regex.
```

Comment:
Can you add an example input and output? I'm still unclear what this does.

Reply:
When we try to match '**' in the text we need the tell regex to match the exact '**' instead of the actual usage of '*' which is zero or more of the previous expression. For that we need to add '\'before '*'. Similarly for all the special characters.

Reply:
So you're escaping special characters in order for the rules to not catch them?